packages/jsii/lib/jsii-diagnostic.ts

```diff
@@ -0,0 +1,791 @@
+    readonly #defaultCategory: ts.DiagnosticCategory;
+
+    // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+    #category?: ts.DiagnosticCategory;
+
+    // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+    #formatter: T;
+
+    private constructor(
+      public readonly code: number,
+      public readonly name: string,
+      category: ts.DiagnosticCategory,
+      formatter: T,
+    ) {
+      this.#defaultCategory = category;
+      this.#formatter = formatter;
+
+      if (code in Code.byCode) {
+        throw new Error(
+          `Attempted to create two instances of ${this.constructor.name} with code ${code}`,
+        );
+      }
+      if (name in Code.byName) {
+        throw new Error(
+          `Attempted to create two instances of ${this.constructor.name} with name ${name}`,
+        );
+      }
+      Code.byCode[code] = Code.byName[name] = this;
+    }
+
+    /**
+     * Determines whether this diagnostic is a compilation error. Diagnostics
+     * where this is `true` cannot have their `category` overridden to a lower
+     * category.
+     */
+    public get isError(): boolean {
+      return this.#defaultCategory === ts.DiagnosticCategory.Error;
+    }
+
+    /**
+     * The diagnostic category this particular code is filed as.
+     */
+    public get category(): ts.DiagnosticCategory {
+      return this.#category ?? this.#defaultCategory;
+    }
+
+    /**
+     * Update the diagnostic category for this particular code. If `isError` is
+     * `true`, attempting to set anything other than `ts.DiagnosticCategory.Error`
+     * will result in an error being throw.
+     *
+     * @param newValue the new diagnostic category to be used.
+     */
+    public set category(newValue: ts.DiagnosticCategory) {
+      if (this.isError && newValue !== ts.DiagnosticCategory.Error) {
+        throw new Error(
+          `Illegal attempt to override category of error ${this.code} to ${ts.DiagnosticCategory[newValue]}`,
+        );
+      }
+      this.#category = newValue;
+    }
+
+    /**
+     * Creates a new `JsiiDiagnostic` message without any source code location
+     * data.
+     *
+     * @param args the arguments to the message formatter.
+     *
+     * @deprecated It is preferred to specify a source code location for problem
+     *             markers. Prefer the use of `create` while providing a value
+     *             for the `location` parameter whenever possible.
+     */
+    public createDetached(...args: Parameters<T>): JsiiDiagnostic {
+      return new JsiiDiagnostic(this, this.#formatter(...args));
+    }
+
+    /**
+     * Creates a new `JsiiDiagnostic` message with source code location denoted
+     * by the provided `location` node.
+     *
+     * @param location the source code location attachment of the message.
+     * @param args     the arguments to the message formatter.
+     */
+    public create(
```

Comment:
Why not 2 orthogonal methods?

```ts
Code.create(...).attach(...);
```

Reply:
I actually want to eventually __require__ all diagnostic messages have a source location attached, so I mean for the node to be provided upfront, always... This is why the other method is deprecated.